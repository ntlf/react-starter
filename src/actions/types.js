export const LOAD_TODOS = 'LOAD_TODOS';
