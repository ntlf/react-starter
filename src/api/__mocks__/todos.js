export const getTodos = async () => [
  {
    id: '1',
    text: 'One'
  },
  {
    id: '2',
    text: 'Two'
  }
];
