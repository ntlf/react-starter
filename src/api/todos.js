import delay from './delay';

export const getTodos = async () => {
  // TODO network

  await delay(1000);

  return [
    {
      id: '1',
      text: 'One'
    },
    {
      id: '2',
      text: 'Two'
    }
  ];
};
