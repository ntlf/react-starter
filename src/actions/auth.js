import { createAction } from 'redux-starter-kit';
import { postLogin } from '../api/auth';

export const loggedIn = createAction('LOGGED_IN');

export const logIn = (username, password) => async dispatch => {
  const token = await postLogin(username, password);
  dispatch(loggedIn(token));
};

export const loggedOut = createAction('LOGGED_OUT');

export const logOut = () => async dispatch => {
  dispatch(loggedOut());
};
