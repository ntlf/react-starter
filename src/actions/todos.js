import { LOAD_TODOS } from './types';
import { getTodos } from '../api/todos';

export const loadTodos = () => async dispatch => {
  const todos = await getTodos();

  dispatch({
    type: LOAD_TODOS,
    payload: todos
  });
};
