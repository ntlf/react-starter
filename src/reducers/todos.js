import { LOAD_TODOS } from '../actions/types';

const initialState = {
  data: []
};

export default (state = initialState, action) => {
  switch (action.type) {
    case LOAD_TODOS:
      return {
        ...state,
        data: [...action.payload]
      };
    default:
      return state;
  }
};
