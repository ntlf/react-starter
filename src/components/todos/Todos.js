import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import { loadTodos } from '../../actions/todos';
import Content from '../common/Content';

export const Todos = ({ todos, loadTodos }) => {
  useEffect(() => {
    loadTodos();
  }, [loadTodos]);

  return (
    <Content>
      <div>Todos</div>

      {!todos.data.length && <div>Loading...</div>}

      {!!todos.data.length && (
        <div>
          <ul>
            {todos.data.map(todo => (
              <li key={todo.id}>{todo.text}</li>
            ))}
          </ul>
        </div>
      )}
    </Content>
  );
};

const mapStateToProps = ({ todos }) => ({
  todos,
});

const mapDispatchToProps = {
  loadTodos,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Todos);
