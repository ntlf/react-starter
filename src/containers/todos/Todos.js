/* eslint-disable react/button-has-type */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { addTodo, loadTodos } from '../../actions/todos';
import styles from './Todos.module.scss';

export class Todos extends Component {
  state = { value: '' };

  componentDidMount() {
    const { loadTodos } = this.props;

    loadTodos();
  }

  handleChange = event => {
    this.setState({ value: event.target.value });
  };

  handleSubmit = event => {
    const { addTodo } = this.props;
    const { value } = this.state;

    addTodo({
      id: Math.random()
        .toString(36)
        .substring(2, 8),
      text: value
    });

    this.setState({ value: '' });

    event.preventDefault();
  };

  render() {
    const { todos } = this.props;
    const { value } = this.state;

    if (todos.error) {
      return <div>{todos.error}</div>;
    }

    return (
      <div className={styles.Home}>
        <form onSubmit={this.handleSubmit}>
          <input type="text" value={value} onChange={this.handleChange} />
          <button>Send</button>
        </form>

        {todos.isFetching && <div>Loading...</div>}
        {todos.data &&
          !todos.isFetching && (
            <div>
              <ul>
                {todos.data.map(todo => (
                  <li key={todo.id}>{todo.text}</li>
                ))}
              </ul>
            </div>
          )}
      </div>
    );
  }
}

Todos.propTypes = {
  loadTodos: PropTypes.func.isRequired,
  addTodo: PropTypes.func.isRequired,
  todos: PropTypes.shape({
    data: PropTypes.array
  })
};

const mapStateToProps = ({ todos }) => ({
  todos
});

const mapDispatchToProps = {
  loadTodos,
  addTodo
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Todos);
