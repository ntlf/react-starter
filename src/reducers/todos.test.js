import todos from './todos';
import { LOAD_TODOS } from '../actions/types';

describe('todos reducer', () => {
  it('is a function', () => {
    expect(todos.constructor).toBe(Function);
  });

  it('should output todos', () => {
    const data = [
      {
        id: '1',
        text: 'One'
      },
      {
        id: '2',
        text: 'Two'
      }
    ];

    const action = {
      type: LOAD_TODOS,
      payload: data
    };
    expect(todos({}, action)).toMatchSnapshot();
  });
});
