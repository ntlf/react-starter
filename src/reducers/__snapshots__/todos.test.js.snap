// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos reducer should output todos 1`] = `
Object {
  "data": Array [
    Object {
      "id": "1",
      "text": "One",
    },
    Object {
      "id": "2",
      "text": "Two",
    },
  ],
}
`;
